Unit 4 - CLIP and relatives - Losses - Intro (#146)
johko/computer-vision-course: chapters/en/Unit 4 - Mulitmodal Models/CLIP and relatives/losses.mdx

# Contrastive Learning

> Before diving into the diffrent losses used to train models like CLIP, it is important to have a clear understanding about what contrastive learning is. Contrastive Learning is an unsupervised deep learning method aimed at representation learning. Its objective is to develop a data representation where similar items are positioned closely in the representation space, and dissimilar items are distinctly separated.

Reviewer: ```suggestion
Before diving into the different losses used to train models like CLIP, it is important to have a clear understanding of what contrastive learning is. Contrastive Learning is an unsupervised deep learning method aimed at representation learning. Its objective is to develop a data representation where similar items are positioned closely in the representation space, and dissimilar items are distinctly separated.
```

> Before diving into the diffrent losses used to train models like CLIP, it is important to have a clear understanding about what contrastive learning is. Contrastive Learning is an unsupervised deep learning method aimed at representation learning. Its objective is to develop a data representation where similar items are positioned closely in the representation space, and dissimilar items are distinctly separated.

Reviewer 2: ```suggestion
Before diving into the diffrent losses used to train models like CLIP, it is important to have a clear understanding about what contrastive learning is. Contrastive Learning is an unsupervised deep learning method aimed at representation learning. Its objective is to develop a data representation where similar items are positioned closely in the representation space and dissimilar items are distinctly separated.
```

> In image below, we have an example where we want to keep the representation from dogs closer to other dogs, but also far from cats.

Reviewer: ```suggestion
In the image below, we have an example where we want to keep the representation from dogs closer to other dogs, but also far from cats.
```

> ![image info](./contrastive_learning.png)

Reviewer 2: You can put your images in hf-vision/course-assets hugging face repository and put the image from here

# Losses

## Contrastive loss

> Contrastive loss is one of the first training objectives that was used for contrastive learning. It takes as input a PAIR of samples that can be similar or dissimilar, and the objective is to bring similar samples closes and dissimilar distant.

Reviewer 3: ```suggestion
Contrastive loss is one of the first training objectives that was used for contrastive learning. It takes as input a pair of samples that can be similar or dissimilar, and the objective is to bring similar samples closes and dissimilar distant.
```

Author: I just wanted to emphasize that this loss is pair wise haha but I already made it lower case

> Contrastive loss is one of the first training objectives that was used for contrastive learning. It takes as input a PAIR of samples that can be similar or dissimilar, and the objective is to bring similar samples closes and dissimilar distant.

Reviewer: ```suggestion
Contrastive loss is one of the first training objectives that was used for contrastive learning. It takes as input a PAIR of samples that can be similar or dissimilar, and the objective is to bring similar samples close and dissimilar distant.
```

> Contrastive loss is one of the first training objectives that was used for contrastive learning. It takes as input a PAIR of samples that can be similar or dissimilar, and the objective is to bring similar samples closes and dissimilar distant.

Reviewer 2: ```suggestion
Contrastive loss is one of the first training objectives used for contrastive learning. It takes a pair of samples as input that can be similar or dissimilar, and the objective is to bring similar samples closes and dissimilar distant.
```

> Contrastive loss is one of the first training objectives that was used for contrastive learning. It takes as input a PAIR of samples that can be similar or dissimilar, and the objective is to bring similar samples closes and dissimilar distant.

Reviewer 4: Perhaps, the sentence needs rephrasing. E.g.:

"Contrastive loss is one of the first training objectives used for contrastive learning. It takes a pair of samples as input that can be similar or dissimilar, and the objective is to map similar samples close in the embedding space and push dissimilar samples apart."

> Technically speaking, imagine that we have a list of input samples $\{x_n\}$ from multiple classes. What we want is a function where examples from the same class have their embeddings close in the embedding space, and also examples from different classes are far apart. Translating this to a mathematical equation, what we have is:

Reviewer 2: ```suggestion
Technically speaking, imagine that we have a list of input samples $\{x_n\}$ from multiple classes. We want a function where examples from the same class have their embeddings close in the embedding space, and examples from different classes are far apart. Translating this to a mathematical equation, what we have is:
```

$$L = \mathbb{1}[y_i = y_j]||x_i - x_j||^2 + \mathbb{1}[y_i \neq y_j]max(0, \epsilon - ||x_i - x_j||^2)$$

Explaining in simple terms:

- If the samples are similar ($y_i = y_j$), then we minimize the term $||x_i - x_j||^2$ that corresponds to their Euclidean distance, i.e., we want to make them closer;
- If the samples are dissimilar ($y_i \neq y_j$), then we minimize the term $max(0, \epsilon - ||x_i - x_j||^2)$ that is equivalent to maximizing their euclidean distance until some limit $\epsilon$, i.e., we want to make them distant from each other.

> # References

Reviewer: ```suggestion
## References
```

> - https://lilianweng.github.io/posts/2021-05-31-contrastive/
> - https://www.baeldung.com/cs/contrastive-learning

Reviewer 3: ```suggestion
- [An Introduction to Contrastive Learning](https://www.baeldung.com/cs/contrastive-learning)
- [Contrastive Representation Learning](https://lilianweng.github.io/posts/2021-05-31-contrastive/)
```
